include/alpaka/block/shared/st/detail/BlockSharedMemStMemberImpl.hpp

```diff
@@ -0,0 +1,168 @@
+                uint32_t off = 0;
+
+                // Iterate over allocated data only
+                while(off < m_allocdBytes)
+                {
+                    // Adjust offset to be aligned
+                    uint32_t const alignedMetaDataOffset
+                        = varChunkEnd<MetaData>(off) - static_cast<uint32_t>(sizeof(MetaData));
+                    ALPAKA_ASSERT_OFFLOAD(
+                        (alignedMetaDataOffset + static_cast<uint32_t>(sizeof(MetaData))) <= m_allocdBytes);
+                    MetaData* metaDataPtr = reinterpret_cast<MetaData*>(m_mem + alignedMetaDataOffset);
+                    off = metaDataPtr->offset;
+
+                    if(metaDataPtr->id == id)
+                        break;
+                }
+
+                // Variable not found.
+                if(off >= m_allocdBytes)
+                    return nullptr;
+
+                return reinterpret_cast<T*>(&m_mem[off - sizeof(T)]);
+            }
+
+            //! Get last allocated variable.
+            template<typename T>
+            T* getLatestVarPtr() const
+            {
+                return reinterpret_cast<T*>(&m_mem[m_allocdBytes - sizeof(T)]);
+            }
+
+        private:
+#if BOOST_COMP_GNUC
+#    pragma GCC diagnostic pop
+#endif
+
+            //! Byte offset to the end of the memory chunk
+            //!
+            //! Calculate bytes required to store a type with a aligned starting offset.
+            //! Start offset to the origin of the user data chunk can be calculated with `result - sizeof(T)`.
+            //! The padding is always before the origin of the user data chunk and can be zero byte.
+            //!
+            //! \tparam T type should fit into the chunk
+            //! \param byteOffset Current byte offset.
+            //! \result Byte offset to the end of the data chunk, relative to m_mem..
+            template<typename T>
+            std::uint32_t varChunkEnd(uint32_t byteOffset) const
+            {
+                static_assert(
+                    core::vectorization::defaultAlignment >= alignof(T),
+                    "Unable to get block shared static memory for types with alignment higher than defaultAlignment!");
+                constexpr std::uint32_t align = static_cast<std::uint32_t>(std::max(TMinDataAlignBytes, alignof(T)));
+                return (byteOffset / align + (byteOffset % align > 0u)) * align + static_cast<uint32_t>(sizeof(T));
+            }
+
+            //! Offset in bytes relative to m_mem to next free data area.
+            //! The last aligned before the free area is always a meta data header.
+            mutable std::uint32_t m_allocdBytes = 0u;
+
+            //! Memory layout
+            //! |Header|Padding|Variable|Padding|Header|....uninitialized Data ....
+            //! Size of padding can be zero if data after padding is already aligned.
+            mutable uint8_t* m_mem;
```

Comment:
I wonder if the implementation would be easier, if we stored the `MetaData` in a separate array. Thus no alignment would be needed for `MetaData`s and finding variables by id could be faster.

Reply:
We do not know how many headers we need at compile time and we do not want to block more shared memory than from being used by the kernel than absolutely necessary.

Reply:
IIUC these are CPU implementations, so it should not be an issue to have a separate 1KiB block for `MetaData`s. That alone would allow 128 shared memory variable declarations in a kernel. I think that could simplify the code a lot, but it is just a suggestion.

Reply:
I prefer the integrated dictionary. Adding 1KiB will introduce another implicit restriction. And will only allow holding 85 variables. (4 byte id and 8 byte pointer)

Reply:
`BlockSharedMemStMember` was, and `BlockSharedMemStMemberImpl` now is, primarily intended to be used  by the OpenMP5 and OpenACC backends, i.e. ultimately on GPU. They have replaced the old CPU shared memory implementations because they avoid malloc, improve memory locality (creating a chunk of memory which might just end up in L1) and limit the amount of shared memory per block to the same amount as on GPU, thereby forcing improved portability.

Reply:
> I prefer the integrated dictionary.

Fine, keep it then!

> Adding 1KiB will introduce another implicit restriction. And will only allow holding 85 variables. (4 byte id and 8 byte pointer)

`MetaData` is defined as:
```c++
struct MetaData {
  uint32_t id = std::numeric_limits<uint32_t>::max();
  uint32_t offset = 0;
};
```
that makes it 8 bytes in size. 1KiB / 8B = 128.

> BlockSharedMemStMember was, and BlockSharedMemStMemberImpl now is, primarily intended to be used by the OpenMP5 and OpenACC backends, i.e. ultimately on GPU. They have replaced the old CPU shared memory implementations.

If they replaced all CPU implementations, then they *are* intended for those as well. But they new implementation is definitely better than any `malloc()`-based one (I did not know the old implemenation). Locality is rightfully important!

Reply:
> If they replaced all CPU implementations, then they are intended for those as well.

Indeed, they are now. Just wanted to point out their "historical" reason for being. They have to keep working in that context as well as possible.

Reply:
> that makes it 8 bytes in size. 1KiB / 8B = 128.

Yes, you are right, I was thinking about an id and pointer instead of keeping the current format. 
Never the less I will keep inlined dictionary.